kvdb/postgres/readwrite_bucket.go

```diff
@@ -0,0 +1,441 @@
+	err := b.tx.QueryRow(
+		"SELECT value FROM "+b.table+" WHERE "+parentSelector(b.id)+
+			" AND key=$1",
+		key,
+	).Scan(&value)
+
+	switch {
+	case err == sql.ErrNoRows:
+		return nil
+
+	case err != nil:
+		panic(err)
+	}
+
+	return *value
+}
+
+func (b *readWriteBucket) ReadCursor() walletdb.ReadCursor {
+	return newReadWriteCursor(b)
+}
+
+// NestedReadWriteBucket retrieves a nested bucket with the given key.
+// Returns nil if the bucket does not exist.
+func (b *readWriteBucket) NestedReadWriteBucket(
+	key []byte) walletdb.ReadWriteBucket {
+
+	if len(key) == 0 {
+		return nil
+	}
+
+	var id int64
+	err := b.tx.QueryRow(
+		"SELECT id FROM "+b.table+" WHERE "+parentSelector(b.id)+
+			" AND key=$1 AND value IS NULL",
+		key,
+	).Scan(&id)
+
+	switch {
+	case err == sql.ErrNoRows:
+		return nil
+
+	case err != nil:
+		panic(err)
+	}
+
+	return newReadWriteBucket(b.tx, b.table, &id)
+}
+
+// CreateBucket creates and returns a new nested bucket with the given key.
+// Returns ErrBucketExists if the bucket already exists, ErrBucketNameRequired
+// if the key is empty, or ErrIncompatibleValue if the key value is otherwise
+// invalid for the particular database implementation.  Other errors are
+// possible depending on the implementation.
+func (b *readWriteBucket) CreateBucket(key []byte) (
+	walletdb.ReadWriteBucket, error) {
+
+	if len(key) == 0 {
+		return nil, walletdb.ErrBucketNameRequired
+	}
+
+	// Check to see if the bucket already exists.
+	var (
+		value *[]byte
+		id    int64
+	)
+	err := b.tx.QueryRow(
+		"SELECT id,value FROM "+b.table+" WHERE "+parentSelector(b.id)+
+			" AND key=$1",
+		key).Scan(&id, &value)
+
+	switch {
+	case err == sql.ErrNoRows:
+
+	case err == nil && value == nil:
+		return nil, walletdb.ErrBucketExists
+
+	case err == nil && value != nil:
+		return nil, walletdb.ErrIncompatibleValue
+
+	case err != nil:
+		return nil, err
+	}
+
+	// Bucket does not yet exist, so create it. Postgres will generate a
+	// bucket id for the new bucket.
+	err = b.tx.QueryRow(
+		"INSERT INTO "+b.table+" (parent_id, key) "+
+			"VALUES($1, $2) RETURNING id",
+		b.id, key,
+	).Scan(&id)
+	if err != nil {
+		return nil, err
+	}
+
+	return newReadWriteBucket(b.tx, b.table, &id), nil
+}
+
+// CreateBucketIfNotExists creates and returns a new nested bucket with
+// the given key if it does not already exist.  Returns
+// ErrBucketNameRequired if the key is empty or ErrIncompatibleValue
+// if the key value is otherwise invalid for the particular database
+// backend.  Other errors are possible depending on the implementation.
+func (b *readWriteBucket) CreateBucketIfNotExists(key []byte) (
+	walletdb.ReadWriteBucket, error) {
+
+	if len(key) == 0 {
+		return nil, walletdb.ErrBucketNameRequired
+	}
+
+	// Check to see if the bucket already exists.
+	var (
+		value *[]byte
+		id    int64
+	)
+	err := b.tx.QueryRow(
+		"SELECT id,value FROM "+b.table+" WHERE "+parentSelector(b.id)+
+			" AND key=$1",
+		key).Scan(&id, &value)
```

Comment:
nit: wrap `.Scan` to new line here and in previous method (for consistency and better readability).

Reply:
fixed